File: day01/sonar_sweep.py
import sys

def count_larger(a):
    cnt = 0
    for i in range(1, len(a)):
        if a[i] > a[i-1]:
            cnt += 1
    return cnt

with open(sys.argv[1]) as f:
    depth = [int(line) for line in f.readlines()]

print('Part 1:', count_larger(depth))

# compute sliding windows
windows = [sum(depth[i:i+3]) for i in range(0, len(depth)-2)]
print('Part 2:', count_larger(windows))

File: day01/sonar_sweep.cpp
#include <stdlib.h>
#include <iostream>
#include <fstream>
#include <vector>

using namespace std;

int num_larger(vector<int> &a) {
    int cnt = 0;
    for (size_t i = 1; i < a.size(); i++)
        if (a[i] > a[i-1])
            cnt++;
    return cnt;
}

int main(int argc, char *argv[]) {
    const string fname = argv[1];

    ifstream infile(fname);
    if (!infile) {
        perror(fname.c_str());
        exit(1);
    }

    vector<int> depth;
    int d;
    while (infile >> d)
        depth.push_back(d);
    
    cout << "Part 1: " << num_larger(depth) << endl;

    // compute sliding windows
    vector<int> windows;
    for (size_t i = 0; i < depth.size() - 2; i++)
        windows.push_back(depth[i] + depth[i+1] + depth[i+2]);
    cout << "Part 2: " << num_larger(windows) << endl;
}
